tests/phpunit/integration/Core/Util/Migration_1_8_1Test.php

```diff
@@ -0,0 +1,249 @@
+	 */
+	protected $migration;
+
+	private $temp_received_identifiers;
+
+	public function setUp() {
+		parent::setUp();
+
+		$this->context        = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$this->options        = new Options( $this->context );
+		$this->user_options   = new User_Options( $this->context );
+		$this->authentication = new Authentication( $this->context, $this->options, $this->user_options );
+		$this->migration      = new Migration_1_8_1( $this->context, $this->options, $this->user_options, $this->authentication );
+
+		$this->temp_received_identifiers = null;
+	}
+
+	public function test_register() {
+		remove_all_actions( 'admin_init' );
+		$this->migration->register();
+
+		$this->assertTrue( has_action( 'admin_init' ) );
+	}
+
+	public function test_migrate_without_using_proxy() {
+		$this->fake_site_connection();
+		$this->migration->migrate();
+		$this->assertNotEquals( Migration_1_8_1::DB_VERSION, $this->get_db_version() );
+	}
+
+	public function test_migrate_without_credentials() {
+		$this->migration->migrate();
+		$this->assertNotEquals( Migration_1_8_1::DB_VERSION, $this->get_db_version() );
+	}
+
+	public function test_migrate_without_users() {
+		$this->fake_proxy_site_connection();
+
+		$this->migration->migrate();
+		$this->assertEquals( Migration_1_8_1::DB_VERSION, $this->get_db_version() );
+	}
+
+	public function test_migrate_without_problem_cases() {
+		$this->create_user( 'administrator', self::FILE_VERIFICATION );
+		$this->create_user( 'subscriber', false );
+
+		$this->fake_proxy_site_connection();
+
+		$this->migration->migrate();
+		$this->assertEquals( Migration_1_8_1::DB_VERSION, $this->get_db_version() );
+	}
+
+	public function test_migrate_with_problem_cases() {
+		$this->create_user( 'administrator', self::FILE_VERIFICATION );
+		$this->create_user( 'subscriber', false );
+		$this->create_user( 'subscriber', false );
+		$this->create_user( 'administrator', false );
+
+		$problem_cases = array(
+			$this->create_user( 'subscriber', self::META_VERIFICATION ),
+			$this->create_user( 'editor', self::FILE_VERIFICATION ),
+		);
+
+		// Set Google profile data for one of the problematic users (see below).
+		$profile = new Profile( new User_Options( $this->context, $problem_cases[0]->ID ) );
+		$profile->set(
+			array(
+				'email' => 'wapuu.wordpress@gmail.com',
+				'photo' => 'https://wapu.us/wp-content/uploads/2017/11/WapuuFinal-100x138.png',
+			)
+		);
+
+		// Ensure the functionality uses Google profile email preferably, but
+		// falls back to WordPress user email if Google profile data not
+		// present.
+		$expected_identifiers = array(
+			'wapuu.wordpress@gmail.com',
+			$problem_cases[1]->user_email,
+		);
+
+		$this->fake_proxy_site_connection();
+		$this->stub_mark_notifications_response();
+
+		$this->migration->migrate();
+		$this->assertReceivedIdentifiers( $expected_identifiers );
+		foreach ( $problem_cases as $user ) {
+			$this->assertNoSiteKitUserOptions( $user->ID );
+		}
```

Comment:
1. We should also assert that the non-problem users' data is untouched
1. Nitpick: let's stick with the PHPUnit convention for naming negated assertions using "Not" with the positive form e.g.  
    - `assertUserHasSiteKitOptions`
    - `assertUserNotHasSiteKitOptions`
  Maybe these should be most specific to `hasVerification` and `hasAccessToken` or `hasConnection`? 

This assertion is fine with me though if it means the user has/not-has any Site Kit user options.

Reply:
Good catch on testing non-problem users' data. Regarding the scope, I think the current approach of checking for any Site Kit user options makes most sense since that's what the logic should do (effectively disconnect the user).